File: frontend/src/lib/sessionSecretGate.test.ts
// Unit-Tests fuer Issue #2139: Session-Secret Fail-Fast (Frontend-Seite).
//
// Spec: docs/specs/bugfix/session_secret_failfast.md (AC-3, AC-4)
//
// Ausfuehrung:
//   cd frontend && node --experimental-strip-types --test \
//     src/lib/sessionSecretGate.test.ts

import { test } from 'node:test';
import assert from 'node:assert/strict';

import { assertSessionSecretConfigured } from './sessionSecretGate.ts';

test('assertSessionSecretConfigured: undefined secret throws', () => {
	assert.throws(() => assertSessionSecretConfigured(undefined));
});

test('assertSessionSecretConfigured: default literal throws', () => {
	assert.throws(() => assertSessionSecretConfigured('dev-secret-change-me'));
});

test('assertSessionSecretConfigured: shorter than 32 chars throws', () => {
	assert.throws(() => assertSessionSecretConfigured('kurz'));
});

test('assertSessionSecretConfigured: exactly 32 chars does not throw', () => {
	assert.doesNotThrow(() => assertSessionSecretConfigured('abcdefghijklmnopqrstuvwxyz012345'));
});

test('assertSessionSecretConfigured: exactly 31 chars throws', () => {
	assert.throws(() => assertSessionSecretConfigured('abcdefghijklmnopqrstuvwxyz01234'));
});

test('assertSessionSecretConfigured: valid 40-char secret does not throw', () => {
	assert.doesNotThrow(() => assertSessionSecretConfigured('a-genuinely-random-forty-char-secret-12'));
});

test('assertSessionSecretConfigured: testFixtureDir exception allows missing secret', () => {
	assert.doesNotThrow(() => assertSessionSecretConfigured(undefined, 'fixtures/openmeteo'));
});

test('assertSessionSecretConfigured: without testFixtureDir the same secret still throws', () => {
	assert.throws(() => assertSessionSecretConfigured(undefined, ''));
});

File: internal/config/session_secret_gate_test.go
package config

import "testing"

func TestValidateSessionSecret_DefaultLiteralFails(t *testing.T) {
	cfg := &Config{SessionSecret: "dev-secret-change-me"}
	if err := ValidateSessionSecret(cfg); err == nil {
		t.Fatal("expected error for unchanged default secret, got nil")
	}
}

func TestValidateSessionSecret_TestFixtureDirException(t *testing.T) {
	cfg := &Config{SessionSecret: "", TestFixtureDir: "/tmp/fixtures"}
	if err := ValidateSessionSecret(cfg); err != nil {
		t.Fatalf("expected nil for TestFixtureDir exception, got: %v", err)
	}
}

func TestValidateSessionSecret_TooShortFails(t *testing.T) {
	cfg := &Config{SessionSecret: "kurz"}
	if err := ValidateSessionSecret(cfg); err == nil {
		t.Fatal("expected error for secret shorter than 32 chars, got nil")
	}
}

func TestValidateSessionSecret_ExactlyMinLengthPasses(t *testing.T) {
	cfg := &Config{SessionSecret: "abcdefghijklmnopqrstuvwxyz012345"} // 32 Zeichen
	if err := ValidateSessionSecret(cfg); err != nil {
		t.Fatalf("expected nil for secret of exactly 32 chars, got: %v", err)
	}
}

func TestValidateSessionSecret_OneBelowMinLengthFails(t *testing.T) {
	cfg := &Config{SessionSecret: "abcdefghijklmnopqrstuvwxyz01234"} // 31 Zeichen
	if err := ValidateSessionSecret(cfg); err == nil {
		t.Fatal("expected error for secret of 31 chars, got nil")
	}
}

func TestValidateSessionSecret_ValidSecretPasses(t *testing.T) {
	cfg := &Config{SessionSecret: "a-genuinely-random-forty-char-secret-1234"}
	if err := ValidateSessionSecret(cfg); err != nil {
		t.Fatalf("expected nil for valid 40-char secret, got: %v", err)
	}
}
